### PR TITLE
Add bandgap benchmark example

### PR DESCRIPTION
**Pull request by wang-hlin**
### Description
Integrate the band-gap benchmark example into PyKale:

1. Add a new folder: `examples/bandgap_benchmark/`.
2. Expose model components for CGCNN, LEFTNet (two variants), and CartNet under `kale.embed.gcn`:
   `CGCNNEncoderLayer`, `CrystalGCN`, `LEFTNetZ`, `LEFTNetProp`, `CartNet`, `GeometricGraphEncoder`, `CartNet_layer`.
3. Create `kale.embed.materials_equivariant` for SE(3)-equivariant components.
4. Add `RegressionTrainer` to `kale.pipeline.base_nn_trainer`.
5. Add metrics/utilities to `kale.evaluate.metrics` (e.g., Gaussian distance expansion for interatomic-distance encoding; mean relative error).

### Status
**Ready**

### Types of changes

- [ ] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [x] In-line docstrings updated.
- [ ] [Source for documentation at `docs`](https://github.com/pykale/pykale/tree/main/docs/source) manually updated for new API.

### CODE REVIEW

File: examples/bandgap_benchmark/configs/configs_by_category/cartnet/config_Antimonides.yaml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  LOG_DIR_NAME: data_by_type/cartnet/Antimonides
+MODEL:
+  CIF_FOLDER: cif_file
+  INIT_FILE: cif_file/atom_init.json
+  NAME: cartnet
+  PRETRAINED_MODEL_PATH: saved_models/cartnet/best.ckpt
+OUTPUT:
+  DIR: predictions/cartnet/Antimonides
+SOLVER:
+  BATCH_SIZE: 64
```
Comment (shuo-zhou) [inline examples/bandgap_benchmark/configs/configs_by_category/cartnet/config_Antimonides.yaml]: Only configure the arguments in `.yaml` if they differ from the defaults specified in `config.py`. Check all `.yaml` files.

File: examples/bandgap_benchmark/main.py [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+        help="gpu id(s) to use. int(0) for cpu. list[x,y] for xth, yth GPU."
+        "str(x) for the first x GPUs. str(-1)/int(-1) for all available GPUs",
+    )
+    parser.add_argument("--resume", default="", type=str)
+    parser.add_argument("--pretrain", action="store_true", help="Enable pretraining mode using the full dataset")
+    args = parser.parse_args()
+    return args
+
+
+def load_data(cfg):
```
Comment (shuo-zhou) [inline examples/bandgap_benchmark/main.py]: To maintain clarity in `main.py`, place custom functions in a separate Python file. If they are reusable, consider adding them to PyKale.”

File: examples/bandgap_benchmark/main.py [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+            atom_fea_len, nbr_fea_len, pos_fea_len = (
+                sample.atom_fea.shape[-1],
+                sample.nbr_fea.shape[-1],
+                sample.positions.shape[-1],
+            )
+
+            if cfg.MODEL.NAME in ["random_forest", "linear_regression", "svm"]:
+                print(f"Training {cfg.MODEL.NAME} model...")
+                X_train, y_train = extract_features(train_dataset)
+                X_val, y_val = extract_features(val_dataset)
```
Comment (shuo-zhou) [inline examples/bandgap_benchmark/main.py]: Rename the varialbes: `x_valid` and `y_valid`. In PyKale’s variable naming convention, avoid using uppercase; use `_valid` for validation data. Please check all.

File: examples/bandgap_benchmark/main.py [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+    # Set random seed for reproducibility
+    set_random_seed(cfg.SOLVER.SEED)
+
+    train_data, val_data = load_data(cfg)
+
+    if args.pretrain:
+        print("Pretraining mode enabled. Splitting training data into k folds for validation.")
+
+        # Split and prepare data
+        kf = KFold(n_splits=cfg.SOLVER.NUM_FOLDS, shuffle=True, random_state=cfg.SOLVER.SEED)
```
Comment (shuo-zhou) [inline examples/bandgap_benchmark/main.py]: Consider moving lines 256–260 into a separate function for clarity and reuse.

File: examples/bandgap_benchmark/README.md
Comment (shuo-zhou) [inline examples/bandgap_benchmark/README.md:1]: Consider adding instructions on the configurable arguments and how to create a .yaml file for configuration.

File: kale/embed/gcn.py [hunk 5]
```diff
@@ -316,3 +331,639 @@ def forward(self, batch_graph):
...
+        super(CGCNNEncoderLayer, self).__init__()
+        self.fc_full = nn.Linear(2 * atom_fea_len + nbr_fea_len, 2 * atom_fea_len)
+        self.sigmoid = nn.Sigmoid()
+        self.softplus1 = nn.Softplus()
+        self.bn1 = nn.BatchNorm1d(2 * atom_fea_len)
+        self.bn2 = nn.BatchNorm1d(atom_fea_len)
+        self.softplus2 = nn.Softplus()
+
+    def forward(self, atom_in_fea, nbr_fea, nbr_fea_idx):
+        N, M = nbr_fea_idx.shape
```
Comment (shuo-zhou) [inline kale/embed/gcn.py]: Avoid using single uppercase letters as variable names

File: kale/embed/gcn.py [hunk 6]
```diff
@@ -316,3 +331,639 @@ def forward(self, batch_graph):
...
+
+    def pooling(self, atom_fea, crystal_atom_idx):
+        """
+        Pooling the atom features to crystal features
+        """
+        summed_fea = [torch.mean(atom_fea[idx_map], dim=0, keepdim=True) for idx_map in crystal_atom_idx]
+        return torch.cat(summed_fea, dim=0)
+
+
+class LEFTNetZ(nn.Module):
```
Comment (shuo-zhou) [inline kale/embed/gcn.py]: Consider moving this under `embed.model_lib`. Let me know if you need a discussion. Also applies to `LEFTNetProp` and `CartNet`.

**Review by shuo-zhou — CHANGES_REQUESTED** [review]
Please add test cases to cover all the changes

File: examples/bandgap_benchmark/configs/configs_by_category/cartnet/config_Antimonides.yaml (hunk 1 above)
Comment (wang-hlin) [reply examples/bandgap_benchmark/configs/configs_by_category/cartnet/config_Antimonides.yaml]: Shall I keep just one YAML file as an example and add detailed instructions in the README? The YAMLs mostly follow the default settings and only differ slightly, mainly in the epoch number and learning rate

**Comment by haipinglu** [comment]
@codex review the full PR diff (all files changed), and report the top 5 Must-fix/Important issues across the whole PR
